src/composition.rs

```diff
@@ -130,8 +130,69 @@ macro_rules! range {
     }
 }
 
-macro_rules! filter {
+macro_rules! binary_eq_expr {
     ($key:expr, $value:expr) => {
+        ast::Expression::Binary(Box::new(ast::BinaryExpr {
+            base: ast::BaseNode::default(),
+            left: ast::Expression::Member(Box::new(
+                ast::MemberExpr {
+                    base: ast::BaseNode::default(),
+                    lbrack: vec![],
+                    rbrack: vec![],
+                    object: ast::Expression::Identifier(
+                        ast::Identifier {
+                            base: ast::BaseNode::default(),
+                            name: "r".into(),
+                        },
+                    ),
+                    property: ast::PropertyKey::Identifier(
+                        ast::Identifier {
+                            base: ast::BaseNode::default(),
+                            name: $key,
+                        },
+                    ),
+                },
+            )),
+            right: ast::Expression::StringLit(ast::StringLit {
+                base: ast::BaseNode::default(),
+                value: $value,
+            }),
+            operator: ast::Operator::EqualOperator,
+        }))
+    };
+}
+
+fn logical_expr(
+    operator: ast::LogicalOperator,
+    key: String,
+    values: &[String],
+) -> Result<ast::Expression, ()> {
+    match values {
+        [] => Err(()),
+        [head] => Ok(binary_eq_expr!(key, head.to_string())),
+        [head, ..] => {
```

Comment:
This is a recursive function, with an unknown length (at compile time) of the values array.

As such, the filter! macro can be compile time (since it only pass the pointer to values.
Then this logical_expr is runtime, based on the runtime recursive depth.
And then the binary_eq_expr! can be a macro.

Reply:
This is the kind of comment that should be part of the function documentation.

Reply:
done.